File: ts/ARC.ts
import { Investigation_$reflection, Investigation } from "./ISA/Investigation.js";
import { FileSystem_$reflection, FileSystem } from "./FileSystem/FileSystem.js";
import { fold } from "./fable_modules/fable-library-ts/Array.js";
import { combine } from "./FileSystem/Path.js";
import { combineAssaySubfolderPaths, combineAssayFolderPath } from "./ARCPath.js";
import { Assay } from "./ISA/Assay.js";
import { Record } from "./fable_modules/fable-library-ts/Types.js";
import { IComparable, IEquatable } from "./fable_modules/fable-library-ts/Util.js";
import { record_type, unit_type, TypeInfo } from "./fable_modules/fable-library-ts/Reflection.js";

export class ARC extends Record implements IEquatable<ARC>, IComparable<ARC> {
    readonly ISA: Investigation;
    readonly CWL: void;
    readonly FileSystem: FileSystem;
    constructor(ISA: Investigation, CWL: void, FileSystem: FileSystem) {
        super();
        this.ISA = ISA;
        this.CWL = CWL;
        this.FileSystem = FileSystem;
    }
    static create(isa: Investigation, cwl: void, fs: FileSystem): ARC {
        return new ARC(isa, void 0, fs);
    }
    static updateISA(isa: Investigation, arc: ARC): ARC {
        throw new Error();
    }
    static updateCWL(cwl: void, arc: ARC): ARC {
        throw new Error();
    }
    static updateFileSystem(fileSystem: FileSystem, arc: ARC): ARC {
        throw new Error();
    }
    static addFile(path: string, arc: ARC): ARC {
        (arg: string): ((arg0: FileSystem) => FileSystem) => ((arg_1: FileSystem): FileSystem => FileSystem.addFile(arg, arg_1));
        (arg_2: FileSystem): ((arg0: ARC) => ARC) => ((arg_3: ARC): ARC => ARC.updateFileSystem(arg_2, arg_3));
        throw new Error();
    }
    static addFolder(path: string, arc: ARC): ARC {
        (arg: string): ((arg0: FileSystem) => FileSystem) => ((arg_1: FileSystem): FileSystem => FileSystem.addFolder(arg, arg_1));
        (arg_2: FileSystem): ((arg0: ARC) => ARC) => ((arg_3: ARC): ARC => ARC.updateFileSystem(arg_2, arg_3));
        throw new Error();
    }
    static addFolders(paths: string[], arc: ARC): ARC {
        fold<string, ARC>((arc_1: ARC, path: string): ARC => ARC.addFolder(path, arc_1), arc, paths);
        throw new Error();
    }
    static addEmptyFolder(path: string, arc: ARC): ARC {
        let arg_2: string;
        (arg: string): ((arg0: FileSystem) => FileSystem) => ((arg_1: FileSystem): FileSystem => FileSystem.addFolder(arg, arg_1));
        (arg_2 = combine(path, ".gitkeep"), (arg_3: FileSystem): FileSystem => FileSystem.addFile(arg_2, arg_3));
        (arg_4: FileSystem): ((arg0: ARC) => ARC) => ((arg_5: ARC): ARC => ARC.updateFileSystem(arg_4, arg_5));
        throw new Error();
    }
    static addEmptyFolders(paths: string[], arc: ARC): ARC {
        fold<string, ARC>((arc_1: ARC, path: string): ARC => ARC.addEmptyFolder(path, arc_1), arc, paths);
        throw new Error();
    }
    static addAssay(assay: Assay, studyIdentifier: string, arc: ARC): ARC {
        const assayFolderPath: string = combineAssayFolderPath(assay);
        const assaySubFolderPaths: string[] = combineAssaySubfolderPaths(assay);
        const assayReadmeFilePath: string = combine(assayFolderPath, "Readme.md");
        const arg_9: Investigation = Investigation.addAssay(assay, studyIdentifier, arc.ISA);
        let arg_10: ARC;
        let arg_8: ARC;
        const arg_6: ARC = ARC.addFolder(assayFolderPath, arc);
        arg_8 = ARC.addEmptyFolders(assaySubFolderPaths, arg_6);
        arg_10 = ARC.addFile(assayReadmeFilePath, arg_8);
        return ARC.updateISA(arg_9, arg_10);
    }
}

export function ARC_$reflection(): TypeInfo {
    return record_type("ARC.ARC", [], ARC, () => [["ISA", Investigation_$reflection()], ["CWL", unit_type], ["FileSystem", FileSystem_$reflection()]]);
}

File: src/ARC/ARC.fs
namespace ARC

open FileSystem
open ISA

open Contract
open Fable.Core

[<AttachMembers>]
type ARC =
    {
       ISA : ISA.Investigation
       CWL : CWL.CWL
       FileSystem : FileSystem.FileSystem 
    }    

    static member create(
        isa: ISA.Investigation,
        cwl: CWL.CWL,
        fs: FileSystem.FileSystem 
    ) =
        {
           ISA = isa
           CWL = cwl
           FileSystem = fs
        }  

    static member updateISA (isa : ISA.Investigation) (arc : ARC) : ARC =
        raise (System.NotImplementedException())

    static member updateCWL (cwl : CWL.CWL) (arc : ARC) : ARC =
        raise (System.NotImplementedException())

    static member updateFileSystem (fileSystem : FileSystem.FileSystem) (arc : ARC) : ARC =
        raise (System.NotImplementedException())

    static member addFile (path : string) (arc : ARC) : ARC =
        FileSystem.addFile |> ignore
        ARC.updateFileSystem |> ignore
        raise (System.NotImplementedException())

    static member addFolder (path : string) (arc : ARC) : ARC =
        FileSystem.addFolder |> ignore
        ARC.updateFileSystem |> ignore
        raise (System.NotImplementedException())

    static member addFolders (paths : string array) (arc : ARC) : ARC =
        paths
        |> Array.fold (fun arc path -> ARC.addFolder path arc) arc |> ignore
        raise (System.NotImplementedException())

    /// Add folder to ARC.FileSystem and add .gitkeep file to the folder
    static member addEmptyFolder (path : string) (arc : ARC) : ARC =
        FileSystem.addFolder  |> ignore   
        FileSystem.addFile (Path.combine path Path.gitKeepFileName) |> ignore
        ARC.updateFileSystem |> ignore
        raise (System.NotImplementedException())

    static member addEmptyFolders (paths : string array) (arc : ARC) : ARC =
        paths
        |> Array.fold (fun arc path -> ARC.addEmptyFolder path arc) arc |> ignore
        raise (System.NotImplementedException())


    /// Add assay folder to the ARC.FileSystem and update the ARC.ISA with the new assay metadata
    static member addAssay (assay : ISA.Assay) (studyIdentifier : string) (arc : ARC) : ARC = 
  
        // - Contracts - //
        // create spreadsheet assays/AssayName/isa.assay.xlsx  
        // create text assays/AssayName/dataset/.gitkeep 
        // create text assays/AssayName/dataset/Readme.md
        // create text assays/AssayName/protocols/.gitkeep 
        // update spreadsheet isa.investigation.xlsx

        // - ISA - //
        let assayFolderPath = Path.combineAssayFolderPath assay
        let assaySubFolderPaths = Path.combineAssaySubfolderPaths assay
        let assayReadmeFilePath = Path.combine assayFolderPath Path.assayReadmeFileName
        let updatedInvestigation = ArcInvestigation.addAssay assay studyIdentifier arc.ISA

        // - FileSystem - //
        // Create assay root folder in ARC.FileSystem
        ARC.addFolder assayFolderPath arc 
        // Create assay subfolders in ARC.FileSystem
        |> ARC.addEmptyFolders assaySubFolderPaths 
        // Create assay readme file in ARC.FileSystem
        |> ARC.addFile assayReadmeFilePath 
        // Update ARC.ISA with the updated investigation
        |> ARC.updateISA updatedInvestigation 

    // to-do: we need a function that generates only create contracts from a ARC data model. 
    // reason: contracts are initially designed to sync disk with in-memory model while working on the arc.
    // but we need a way to create an arc programmatically and then write it to disk.

    // to-do: function that returns read contracts based on a list of paths.
    // the list of paths is used to create a filesystem tree
    static member createReadContracts (paths : string array) : Contract array =
        let fileTree = FileSystemTree.fromFilePaths paths
        raise (System.NotImplementedException())